pkg/api/operators/apis/experiment/v1alpha2/types.go

```diff
@@ -0,0 +1,194 @@
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type ExperimentSpec struct {
+	// List of hyperparameter configurations.
+	Parameters []ParameterSpec `json:"parameters,omitempty"`
+
+	// Describes the objective of the experiment.
+	Objective *ObjectiveSpec `json:"objective,omitempty"`
+
+	// Describes the suggestion algorithm.
+	Algorithm *AlgorithmSpec `json:"algorithm,omitempty"`
+
+	// Template for each run of the trial.
+	TrialRunSpec *TrialRunSpec `json:"trialRunSpec,omitempty"`
+
+	ParallelTrialCount int `json:"parallelTrialCount,omitempty"`
+	MaxIterations      int `json:"maxIterations,omitempty"`
+
+	// TODO - figure out what to do with metric collectors
+	MetricsCollectorType string `json:"metricsCollectorSpec,omitempty"`
+
+	// Other fields, exact format is TBD
+	// - NAS
+	// - early stopping
+	// - resume experiment
+}
+
+type ExperimentStatus struct {
+	StartTime         *metav1.Time `json:"startTime,omitempty"`
+	CompletionTime    *metav1.Time `json:"completionTime,omitempty"`
+	LastReconcileTime *metav1.Time `json:"lastReconcileTime,omitempty"`
+
+	Conditions []ExperimentCondition `json:"conditions,omitempty"`
+
+	// Since a Trial encapsulates suggestion + evaluation + observation, the user
+	// can access all relevant information here.
+	CurrentOptimalTrial *Trial `json:"currentOptimalTrial,omitempty"`
+
+	SuggestionRoundCount int `json:" suggestionRoundCount,omitempty"`
+
+	// Should we include more metadata about Trials here?
+	// E.g. Number of Trials completed/failed/pending
+}
+
+// +k8s:deepcopy-gen=true
+// ExperimentCondition describes the state of the experiment at a certain point.
+type ExperimentCondition struct {
+	// Type of experiment condition.
+	Type ExperimentConditionType `json:"type"`
+
+	// Status of the condition, one of True, False, Unknown.
+	Status v1.ConditionStatus `json:"status"`
+
+	// The reason for the condition's last transition.
+	Reason string `json:"reason,omitempty"`
+
+	// A human readable message indicating details about the transition.
+	Message string `json:"message,omitempty"`
+
+	// The last time this condition was updated.
+	LastUpdateTime metav1.Time `json:"lastUpdateTime,omitempty"`
+
+	// Last time the condition transitioned from one status to another.
+	LastTransitionTime metav1.Time `json:"lastTransitionTime,omitempty"`
+}
+
+// ExperimentConditionType defines the state of an Experiment.
+type ExperimentConditionType string
+
+const (
+	ExperimentCreated    ExperimentConditionType = "Created"
+	ExperimentRunning    ExperimentConditionType = "Running"
+	ExperimentRestarting ExperimentConditionType = "Restarting"
+	ExperimentSucceeded  ExperimentConditionType = "Succeeded"
+	ExperimentFailed     ExperimentConditionType = "Failed"
+)
+
+type ParameterSpec struct {
+	Name          string        `json:"name,omitempty"`
+	ParameterType ParameterType `json:"parameterType,omitempty"`
+	FeasibleSpace FeasibleSpace `json:"feasibleSpace,omitempty"`
```

Comment:
I'd still really like to add the ability to transform parameters here. It's a really common use case (e.g. tuning over the log learning rate form -6 to -3 rather than the actual learning rate from 1e-6 to 1e-3), won't complicate anything else, and easy to implement in the suggestion package.

Reply:
That won't require any API or CRD changes right? If I understand this correctly.

Reply:
It shouldn't. It will only get used inside the suggestion services to transform the parameter space before generating suggestions